src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1821,41 +1925,168 @@ AddTypeConversion(Node *originalAggregate, Node *newExpression)
-	Node *havingQual = originalOpNode->havingQual;
-	AttrNumber targetProjectionNumber = 1;
-	WorkerAggregateWalkerContext *walkerContext =
-		palloc0(sizeof(WorkerAggregateWalkerContext));
 	Index nextSortGroupRefIndex = 0;
-	bool queryHasAggregates = false;
-	bool distinctClauseSupersetofGroupClause = false;
 	bool distinctPreventsLimitPushdown = false;
-	bool createNewGroupByClause = false;
+	bool groupByExtended = false;
 	bool groupedByDisjointPartitionColumn =
 		extendedOpNodeProperties->groupedByDisjointPartitionColumn;
-	bool pushDownWindowFunction = extendedOpNodeProperties->pushDownWindowFunctions;
 
-	walkerContext->expressionList = NIL;
-	walkerContext->pullDistinctColumns = extendedOpNodeProperties->pullDistinctColumns;
+	QueryTargetList queryTargetList;
+	QueryGroupClause queryGroupClause;
+	QueryDistinctClause queryDistinctClause;
+	QueryWindowClause queryWindowClause;
+	QueryOrderByLimit queryOrderByLimit;
+	Node *queryHavingQual = NULL;
 
-	nextSortGroupRefIndex = GetNextSortGroupRef(targetEntryList);
+	List *originalTargetEntryList = originalOpNode->targetList;
+	List *originalGroupClauseList = originalOpNode->groupClauseList;
+	List *originalSortClauseList = originalOpNode->sortClauseList;
+	Node *originalHavingQual = originalOpNode->havingQual;
+	Node *originalLimitCount = originalOpNode->limitCount;
+	Node *originalLimitOffset = originalOpNode->limitOffset;
+	List *originalWindowClause = originalOpNode->windowClause;
+	List *originalDistinctClause = originalOpNode->distinctClause;
+	bool hasDistinctOn = originalOpNode->hasDistinctOn;
+
+	int originalGroupClauseLength = list_length(originalGroupClauseList);
+	bool queryHasAggregates = TargetListHasAggragates(originalTargetEntryList);
+
+	/* calculate the next sort group index based on the original target list */
+	nextSortGroupRefIndex = GetNextSortGroupRef(originalTargetEntryList);
+
+	/* targetProjectionNumber starts from 1 */
+	queryTargetList.targetProjectionNumber = 1;
+
+	/* worker query always include all the group by entries in the query */
+	queryGroupClause.groupClauseList = copyObject(originalGroupClauseList);
+
+	/*
+	 * nextSortGroupRefIndex is used by group by, window and order by clauses.
+	 * Thus, we pass a reference to a single nextSortGroupRefIndex and expect
+	 * it modified separately while processing those parts of the query.
+	 */
+	queryGroupClause.nextSortGroupRefIndex = &nextSortGroupRefIndex;
+	queryWindowClause.nextSortGroupRefIndex = &nextSortGroupRefIndex;
+	queryOrderByLimit.nextSortGroupRefIndex = &nextSortGroupRefIndex;
+
+	/* initialize the remaining parts to default values */
+	queryTargetList.targetEntryList = NIL;
+	queryDistinctClause.workerDistinctClause = NIL;
+	queryDistinctClause.workerHasDistinctOn = false;
+	queryWindowClause.hasWindowFunctions = false;
+	queryWindowClause.workerWindowClauseList = NIL;
+	queryOrderByLimit.workerLimitCount = NULL;
+	queryOrderByLimit.workerSortClauseList = NIL;
```

Comment:
it might be worth just doing a `memset` to initialise to 0 instead, in case we ever forget a field.

Reply:
makes sense